File: problems/0022-generate-parentheses/typescript/solution.ts
// Generate Parentheses — https://leetcode.com/problems/generate-parentheses/
export function generateParenthesis(n: number): string[] {
  const res: string[] = [];
  const backtrack = (current: string, open: number, close: number): void => {
    if (current.length === 2 * n) {
      res.push(current);
      return;
    }
    if (open < n) backtrack(current + "(", open + 1, close);
    if (close < open) backtrack(current + ")", open, close + 1);
  };
  backtrack("", 0, 0);
  return res;
}

File: problems/0022-generate-parentheses/go/solution.go
package solution

// Generate Parentheses — https://leetcode.com/problems/generate-parentheses/
func generateParenthesis(n int) []string {
	res := []string{}
	var backtrack func(cur string, open, closed int)
	backtrack = func(cur string, open, closed int) {
		if len(cur) == 2*n {
			res = append(res, cur)
			return
		}
		if open < n {
			backtrack(cur+"(", open+1, closed)
		}
		if closed < open {
			backtrack(cur+")", open, closed+1)
		}
	}
	backtrack("", 0, 0)
	return res
}
